## 실습 3-2-1. 랜덤 벡터 생성과 정렬
- 길이 10인 랜덤 벡터(1차원 배열)을 생성하고 이를 정렬한다
- 오름차순과 내림차순으로 한 번씩 정렬해 본다

In [ ]:
import numpy as np

In [6]:
arr = np.random.random(10)
arr.sort()

In [7]:
arr # 오름차순 정렬

array([ 0.030492  ,  0.12043689,  0.32015067,  0.33430422,  0.72236358,
        0.83430805,  0.88724013,  0.8975324 ,  0.91277164,  0.98039898])

In [8]:
arr[::-1].sort()

In [9]:
arr # 내림차순 정렬

array([ 0.98039898,  0.91277164,  0.8975324 ,  0.88724013,  0.83430805,
        0.72236358,  0.33430422,  0.32015067,  0.12043689,  0.030492  ])

## 실습 3-2-2. 벡터를 행렬로 변환하기
- 길이 9의 벡터(1차원 배열)를 생성하고 이를 3X3 크기의 행렬로 변환한다
    - NumPy의 arange() 함수를 이용해 0부터 8까지의 정수를 담은 1차원 배열을 생성한다
    - 이를 reshape() 함수로 2차원 배열로 변환한다

In [13]:
arr = np.arange(9)

In [16]:
arr = arr.reshape(3,3)

In [17]:
arr

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

## 실습 3-2-3. 대각행렬(diagonal matrix)
- 실습 3-2-2에서 생성한 행렬을 대각행렬로 변환해 본다
    - 우선 행렬의 주대각 원소(대각선상에 있는 원소)만 1차원 배열로 출력해 본다
    - 3X3  대각행렬(정사각행렬의 주대각 성분 이외의 모든 성분이 0인 행렬)로 변환해 본다

In [20]:
arr = np.arange(9).reshape(3,3)
diag = np.diag(arr)

In [21]:
diag

array([0, 4, 8])

In [22]:
diag_mat = np.diag(np.diag(arr))

In [23]:
diag_mat

array([[0, 0, 0],
       [0, 4, 0],
       [0, 0, 8]])

## 실습 3-2-4. 행렬 정규화(normalization)
- 정수로 이루어진 5X5 행렬을 만들고 이를 정규화해본다
    - randint() 함수를 이용해 1부터 100 사이의 정수로 이루어진 행렬을 생성한다
    - 행렬 내의 값들을 정규화한다

In [32]:
arr = np.random.randint(1, 100, size =(5,5))

In [33]:
arr

array([[ 2, 37,  7, 27, 14],
       [83, 64, 96, 95,  4],
       [26, 12, 44, 78, 33],
       [95, 92, 69, 90, 85],
       [34, 49, 10, 96, 82]])

In [36]:
arrMax, arrMin = arr.max(), arr.min()
arr_normalized = (arr - arrMin)/(arrMax - arrMin)

In [37]:
arr_normalized

array([[ 0.        ,  0.37234043,  0.05319149,  0.26595745,  0.12765957],
       [ 0.86170213,  0.65957447,  1.        ,  0.9893617 ,  0.0212766 ],
       [ 0.25531915,  0.10638298,  0.44680851,  0.80851064,  0.32978723],
       [ 0.9893617 ,  0.95744681,  0.71276596,  0.93617021,  0.88297872],
       [ 0.34042553,  0.5       ,  0.08510638,  1.        ,  0.85106383]])

## 실습 3-2.5. 행렬 곱(matrix multiplication)
- (2X3) 행렬과 (3X2) 행렬 사이의 곱 연산을 수행한다
    - np.dot() 함수를 활용한다

In [41]:
A = np.array([[1,2,3],[4,5,6]])

In [42]:
B = np.array([[1,2], [3,4], [5,6]])

In [43]:
result = np.dot(A, B)

In [44]:
result

array([[22, 28],
       [49, 64]])

## 실습 3-2-6. 주식 데이터 불러오기
- stock-data.csv의 데이터를 불러온다
    - 1열부터 4열(‘Open’, ‘High’, ‘Low’, ‘Volume’)까지의 데이터는 x_data로 저장하고 5열(‘Close’)의 데이터는 y_data로 저장한다
    - 즉, x_data의 shape은 (732,4), y_data의 shape은 (732,) 혹은 (732,1)이 되어야 한다

In [5]:
data = np.loadtxt('stock-data.csv', delimiter = ',')
data.shape

(732, 5)

In [6]:
x_data = data[:,:-1]

(732, 4)

In [7]:
y_data = data[:, -1]

(732,)

In [8]:
x_data.shape

(732, 4)

In [9]:
x_data

array([[  8.28659973e+02,   8.33450012e+02,   8.28349976e+02,
          1.24770000e+06],
       [  8.23020020e+02,   8.28070007e+02,   8.21655029e+02,
          1.59780000e+06],
       [  8.19929993e+02,   8.24400024e+02,   8.18979980e+02,
          1.28170000e+06],
       ..., 
       [  5.66892592e+02,   5.67002574e+02,   5.56932537e+02,
          1.08000000e+04],
       [  5.61202549e+02,   5.66432590e+02,   5.58672539e+02,
          4.12000000e+04],
       [  5.68002570e+02,   5.68002570e+02,   5.52922516e+02,
          1.31000000e+04]])

In [10]:
y_data.shape

(732,)

In [11]:
y_data

array([ 831.659973,  828.070007,  824.159973,  818.97998 ,  820.450012,
        819.23999 ,  813.669983,  809.559998,  808.380005,  806.969971,
        801.340027,  801.48999 ,  798.530029,  795.695007,  796.789978,
        802.320007,  823.309998,  832.150024,  835.669983,  823.869995,
        819.309998,  805.02002 ,  802.174988,  806.070007,  804.609985,
        807.880005,  806.359985,  807.909973,  804.789978,  806.650024,
        806.150024,  794.02002 ,  786.900024,  786.140015,  771.820007,
        782.789978,  785.049988,  791.549988,  789.909973,  791.26001 ,
        794.559998,  796.419983,  794.200012,  790.799988,  797.849976,
        797.070007,  796.099976,  789.27002 ,  789.289978,  776.419983,
        771.190002,  759.109985,  762.52002 ,  750.5     ,  747.919983,
        758.039978,  770.840027,  768.23999 ,  761.679993,  760.98999 ,
        768.27002 ,  769.200012,  760.539978,  771.22998 ,  764.47998 ,
        758.48999 ,  736.080017,  754.02002 ,  762.559998,  785.

## 실습 3-2-7. 당뇨 데이터 불러오기
- diabetes-data.csv의 데이터를 불러온다
    - 1열부터 8열까지의 데이터는 x_data로 저장하고 9열의 데이터는 y_data로 저장한다
    - 즉, x_data의 shape은 (759, 8), y_data의 shape은 (759,) 혹은 (759,1)이 되어야 한다

In [12]:
data = np.loadtxt('diabetes-data.csv', delimiter = ',')
data.shape

(759, 9)

In [13]:
x_data = data[:, :-1]
y_data = data[:, 5:6]

In [14]:
x_data.shape

(759, 8)

In [15]:
x_data

array([[-0.294118  ,  0.487437  ,  0.180328  , ...,  0.00149028,
        -0.53117   , -0.0333333 ],
       [-0.882353  , -0.145729  ,  0.0819672 , ..., -0.207153  ,
        -0.766866  , -0.666667  ],
       [-0.0588235 ,  0.839196  ,  0.0491803 , ..., -0.305514  ,
        -0.492741  , -0.633333  ],
       ..., 
       [-0.411765  ,  0.21608   ,  0.180328  , ..., -0.219076  ,
        -0.857387  , -0.7       ],
       [-0.882353  ,  0.266332  , -0.0163934 , ..., -0.102832  ,
        -0.768574  , -0.133333  ],
       [-0.882353  , -0.0653266 ,  0.147541  , ..., -0.0938897 ,
        -0.797609  , -0.933333  ]])

In [16]:
y_data.shape

(759,)

In [17]:
y_data

array([ 0.,  1.,  0.,  1.,  0.,  1.,  0.,  1.,  0.,  0.,  1.,  0.,  1.,
        0.,  0.,  0.,  0.,  1.,  0.,  1.,  1.,  0.,  0.,  0.,  0.,  1.,
        1.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,  1.,  0.,  0.,  0.,  1.,
        1.,  1.,  0.,  1.,  0.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,  0.,
        1.,  1.,  0.,  1.,  1.,  1.,  1.,  0.,  1.,  1.,  0.,  1.,  0.,
        1.,  1.,  1.,  0.,  1.,  0.,  1.,  1.,  1.,  1.,  1.,  0.,  1.,
        1.,  1.,  1.,  1.,  0.,  1.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,
        0.,  1.,  1.,  1.,  1.,  1.,  0.,  0.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  0.,  0.,  0.,  1.,  1.,  0.,  0.,  0.,  1.,  1.,
        1.,  0.,  1.,  1.,  1.,  0.,  0.,  1.,  1.,  0.,  0.,  0.,  0.,
        0.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  0.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  0.,  1.,  0.,  0.,  1.,  1.,
        1.,  0.,  1.,  1.,  1.,  1.,  0.,  0.,  1.,  1.,  1.,  1.,  0.,
        0.,  1.,  1.,  1.,  0.,  1.,  0.,  1.,  0.,  1.,  1.,  1

In [1]:
import pandas as pd

## 실습 3-2-8. 데이터프레임 생성하기
- 아래 내용의 데이터를 담은 데이터프레임을 생성해 본다
    - 1열부터 8열까지의 데이터는 x_data로 저장하고 9열의 데이터는 y_data로 저장한다
    - Title, Artist, Length는 String 타입으로, Year는 int 타입으로 저장한다

In [2]:
title = ['Skyfall', 'Lose Yourself', 'Another Brick in the Wall', 'Use Somebody', 'Treasure']
artist = ['Adele', 'Eminem', 'Pink Floyd', 'Kings of Leon', 'Bruno Mars']
length = ['4:46', '5:26', '3:59', '3:51', '2:58']
year = [2012, 2002, 1979, 2008, 2013]
data = pd.DataFrame({'Title': title, 'Artist': artist, 'Length': length, 'Year': year})

In [3]:
data

Artist Length                      Title  Year
0          Adele   4:46                    Skyfall  2012
1         Eminem   5:26              Lose Yourself  2002
2     Pink Floyd   3:59  Another Brick in the Wall  1979
3  Kings of Leon   3:51               Use Somebody  2008
4     Bruno Mars   2:58                   Treasure  2013

## 실습 3-2-9. 데이터프레임 나누기
- 실습 3-2-8에서 생성된 데이터프레임에서 ‘Title’ 컬럼의 데이터만 추출한다
    - 추출한 데이터는 모두 소문자(lowercase)로 변환한 뒤 title_lowercase 이름의 리스트에 저장한다

In [4]:
title_lowercase = [t.lower() for t in data['Title']]

In [5]:
title_lowercase

['skyfall',
 'lose yourself',
 'another brick in the wall',
 'use somebody',
 'treasure']

## 실습 3-2-10. 엑셀 데이터 불러오기
- animals.xlsx 파일을 불러와 데이터프레임을 생성한다
    - 엑셀 파일의 첫 세 컬럼(name, hair, feathers)의 데이터만 포함하는 데이터프레임 animals를 생성한다
    - 참고: .loc[] 혹은 .iloc[] 를 통해 슬라이싱한다(refer to: https://stackoverflow.com/questions/10665889/how-to-take-column-slices-of-dataframe-in-pandas) 

In [2]:
data = pd.read_excel('animals.xlsx')

In [ ]:
data

In [3]:
animals = data.loc[:, ['name', 'hair', 'feathers']]

In [ ]:
animals

## 실습 3-2-11. 엑셀 파일로 저장하기
- 실습 3-2-10에서 생성된 animals 데이터프레임의 데이터를 엑셀 파일로 저장한다
    - animals_sub.xlsx 이름으로 저장한다 

In [13]:
animals.to_excel('animals_sub.xlsx')